int32 freq
---
bool done
